-masterdatadir for datadir bootstrapping

kallewoof opened the pull request:
This PR adds the ability to bootstrap another node from an existing datadir.

To fire up a new node with access to an existing node's datadir `y`, you would simply do

```bash
$ bitcoind -datadir=x -masterdatadir=y
```

(presumably permanently in `bitcoin.conf`) and it would use the blocks files from the former up until it needed to change something, at which point it would copy. (See accompanying test.)

It currently only supports `blocks/blk*` caching, which is implemented as copy-on-write. Due to limitations in LevelDB, the `chainstate` and `blocks/index` dirs are copied to `-datadir` on first launch. Latter is only ~75 MB, but former is ~4 GB, so this is sad but better than nothing.

Note: the `feature_masterdatadir.py` test makes ~150 MB of block data. May want to combine this with the pruning tests to not redo the work of mining, but it is in its own test for now. Also may wanna move it to extended, but keeping it in base while this PR is ongoing.

kallewoof committed:
test: basic -masterdatadir tests

kallewoof committed:
util: add master datadir path support

kallewoof committed:
util: add CopyDirectory helper

laanwj commented:
Concept ACK!

promag commented on src/validation.cpp:
```diff
@@ -3787,6 +3788,29 @@ static FILE* OpenDiskFile(const CDiskBlockPos &pos, const char *prefix, bool fRe
+            uint8_t buffer[16384];
+            while (!feof(mfile)) {
+                size_t r = fread(buffer, 1, 16384, mfile);
+                if (r) fwrite(buffer, 1, r, file);
```
Should handle failure instead of ignoring? Could lead to corrupt blocks?

promag commented on test/functional/feature_masterdatadir.py:
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2017 The Bitcoin Core developers
```
2018

promag reviewed (commented):
What if there is a node (A) running in the master data dir of node (B) and A is writing a block at the same time B is reading?

kallewoof committed:
validation: add g_has_master_datadir and g_master_endblock

kallewoof committed:
init: add -masterdatadir arg

kallewoof committed:
init: copy chainstate and blocks/index to datadir if -masterdatadir is used

Also find and track the last blk/rev files that we rely on from master, as we do not want to load anything beyond that, in case master runs ahead of us at some point.

kallewoof committed:
validation: use or copy master blocks file when local datadir file does not exist

kallewoof commented:
Experiment: skip last byte for case "!readonly & copy existing":

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index b5646f107..7af6956b3 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3804,6 +3804,11 @@ static FILE* OpenDiskFile(const CDiskBlockPos &pos, const char *prefix, bool fRe
             while (!feof(mfile)) {
                 size_t r = fread(buffer, 1, 16384, mfile);
                 if (r) {
+                    // TEST: CORRUPTION
+                    if (r > 0 && r < 16384) {
+                        LogPrintf("*** Corrupting file %s by writing %d/%d bytes from master\n", path.string(), r-1, r);
+                        r--;
+                    }
                     if (r != fwrite(buffer, 1, r, file)) {
                         fclose(mfile);
                         fclose(file);
```

Result: slave node crashes with a 'corrupt block database' error:
```
2018-07-26T02:35:14Z Loaded best chain: hashBestChain=00000000000000000029ac6548add211c4f3d20d20da3d5858c3eda1fb84fece height=529951 date=2018-06-30T22:01:25Z progress=0.977126
2018-07-26T02:35:14Z init message: Rewinding blocks...
2018-07-26T02:35:16Z *** Corrupting file /Users/karljohan-alm/workspace/bitcoin-kw/src/slavedata/blocks/blk01301.dat by writing 5744/5745 bytes from master
2018-07-26T02:35:17Z init message: Verifying blocks...
2018-07-26T02:35:17Z Verifying last 6 blocks at level 3
2018-07-26T02:35:17Z [0%]...ERROR: ReadBlockFromDisk: Deserialize or I/O error - CAutoFile::read: end of file: unspecified iostream_category error at CBlockDiskPos(nFile=1301, nPos=97518532)
2018-07-26T02:35:17Z ERROR: VerifyDB(): *** ReadBlockFromDisk failed at 529951, hash=00000000000000000029ac6548add211c4f3d20d20da3d5858c3eda1fb84fece
2018-07-26T02:35:17Z : Corrupted block database detected.
Please restart with -reindex or -reindex-chainstate to recover.
: Corrupted block database detected.
Please restart with -reindex or -reindex-chainstate to recover.
2018-07-26T02:35:17Z Aborted block database rebuild. Exiting.
2018-07-26T02:35:17Z Shutdown: In progress...
2018-07-26T02:35:17Z scheduler thread interrupt
2018-07-26T02:35:17Z Shutdown: done
```

Disabling corruption patch and restarting results in another crash. Only fix was to -reindex, which initiated a full redownload of blocks (master was pruned).

I think it would be interesting to patch the 'corrupt!' logic with a fallback to grab master file again before failing. Actually, I think I'm going to do that.

jonasschnelli commented on src/init.cpp:
```diff
@@ -1256,6 +1260,27 @@ bool AppInitMain()
     }
 
+    if (g_has_master_datadir) {
+        // LDB does not support readonly access, so we need to copy the entire
```
nit: s/LDB/levelDB

jonasschnelli commented on src/init.cpp:
```diff
@@ -1256,6 +1260,27 @@ bool AppInitMain()
+        // LDB does not support readonly access, so we need to copy the entire
+        // chainstate and blocks/index dirs over to the new location, if not found already
+        fs::path path = GetDataDir() / "chainstate";
+        if (!fs::exists(path)) {
```
Doesn't this copy action require a protection that the masterdatadir is currently not in use/write? If a Bitcoin Core instance is using this (master) datadir it may result in a corrupted copy?

kallewoof replied:
Yes. Master is guaranteed to be safe as no write operation is ever done from the slave, but the opposite is not the case. See https://github.com/bitcoin/bitcoin/pull/13746#issuecomment-407957987 for experiment on corruption. There's no good way to recover, as the slave will have already written data and closed the file before realizing it was truncated.

The question remains how frequent this kind of race condition is, though.

jonasschnelli commented:
Concept ACK

jonasschnelli commented:
I'm more worried about hot copying the levelDB part (chainstate/). Imo to be safe, the master must close the leveldb environment.

sipa commented:
You can't do this while an instance is running on the master (unless you can somehow force it to temporarily close its database environment and pause operations). If you don't need it to work while the master is running you should grab the lock on the master to avoid it.

An alternative is to not do any hardlinking, but just permit a second location where blocks can be found (and just blocks, not the databases, which is safe as block files are append only). You'd "copy" by doing a reindex while specifying the other directory as second path.

kallewoof committed:
util: add ObtainDirectoryLock() for temporary dir locks

kallewoof committed:
linter: add exception to locale dependent function linter

kallewoof commented:
@sipa I am now locking master temporarily while doing the copying, and init-error-shutdown if it is already locked.

I am also now tracking the last block file that the slave relies on in master, so it won't accidentally start reading from master block files if master runs ahead in file count (e.g. slave writes blk123, master then creates blk124 and slave starts trying to read from it).

kallewoof commented:
@jonasschnelli 

> I'm more worried about hot copying the levelDB part (chainstate/). Imo to be safe, the master must close the leveldb environment.

This should be addressed now that the master is locked before copy.

@promag 

> What if there is a node (A) running in the master data dir of node (B) and A is writing a block at the same time B is reading?

The slave will now copy the very last block file over on initial start. It also holds a lock on the master dir for the duration of these operations, so your concerns have been addressed, I think.

jonasschnelli commented:
@kallewoof 
Does the lock works against masters not running this change (0.17 as example)?

kallewoof commented:
> Does the lock works against masters not running this change (0.17 as example)?

Yes. The slave locks the master datadir before copying, and after the initial copying is done, it no longer relies on master playing nice. Since the datadir locking is done in older clients as well, this should (1) cause the slave to fail if the master is running when it's started up for the first time, and (2) cause the master to fail to startup if the slave is in the process of copying.

practicalswift commented on test/functional/feature_masterdatadir.py:
```diff
@@ -0,0 +1,112 @@
+        for i in range(150):
+            mine_large_block(self.nodes[0])
+            if i % 50 == 0:
+                self.log.info("... " + str(i) + " / 150")
```
Use placeholders in the log message like this:

```
self.log.info("... %s / 150", i)
```

It is the preferred form since it allows for lazy evaluation.

Applies to all logging calls in this PR :-)

kallewoof replied:
Fixed!

practicalswift commented:
Concept ACK
